File: src/main/java/org/alpacaindustries/iremiaminigamecore/minigame/Minigame.java
package org.alpacaindustries.iremiaminigamecore.minigame;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;

import net.kyori.adventure.text.Component;
import org.bukkit.event.player.PlayerQuitEvent;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.function.Consumer;
import java.util.Objects;

/**
 * Base abstract class for all minigames.
 * Provides the core functionality and lifecycle management for minigames.
 * <p>
 * This class is thread-safe for all public methods unless otherwise noted.
 */
public abstract class Minigame implements Listener {

  private final @NotNull String id;
  private final @NotNull String displayName;
  private final @NotNull MinigameManager manager;
  private final Set<UUID> players = Collections.synchronizedSet(new HashSet<>());
  private final List<Consumer<Minigame>> endListeners = Collections.synchronizedList(new ArrayList<>());
  private final Map<UUID, Player> playerCache = Collections.synchronizedMap(new HashMap<>());

  private volatile Optional<Location> spawnPoint = Optional.empty();
  private volatile MinigameState state = MinigameState.WAITING;
  private volatile int minPlayers;
  private volatile int maxPlayers;
  private volatile boolean allowJoinDuringGame = false;
  private boolean shouldLoop = false;
  private long loopDelayTicks = 60L; // Default: 3 seconds (20 ticks = 1 second)

  /**
   * Create a new minigame.
   *
   * @param id          Unique identifier for this minigame
   * @param displayName User-friendly name for this minigame
   * @param manager     The manager that created this minigame
   * @throws IllegalArgumentException if any parameter is null or invalid
   */
  protected Minigame(@NotNull String id, @NotNull String displayName, @NotNull MinigameManager manager) {
    this.id = Objects.requireNonNull(id, "Minigame ID cannot be null").trim();
    if (this.id.isEmpty()) {
      throw new IllegalArgumentException("Minigame ID cannot be empty");
    }
    this.displayName = Objects.requireNonNull(displayName, "Display name cannot be null").trim();
    if (this.displayName.isEmpty()) {
      throw new IllegalArgumentException("Display name cannot be empty");
    }
    this.manager = Objects.requireNonNull(manager, "MinigameManager cannot be null");
    this.minPlayers = MinigameConfig.getDefaultMinPlayers();
    this.maxPlayers = MinigameConfig.getDefaultMaxPlayers();
  }

  /**
   * Initialize the minigame. Registers event listeners and sets initial state.
   */
  public void initialize() {
    setState(MinigameState.WAITING);
    getManager().getPlugin().getServer().getPluginManager().registerEvents(this, getManager().getPlugin());
  }

  /**
   * Start the minigame. Only allowed from WAITING or COUNTDOWN state.
   */
  public synchronized void start() {
    if (state != MinigameState.WAITING && state != MinigameState.COUNTDOWN) {
      return;
    }
    if (players.size() < minPlayers) {
      broadcastMessage(Component.text(MinigameConfig.getMsgNotEnoughPlayers()));
      return;
    }
    setState(MinigameState.RUNNING);
    onStart();
  }

  /**
   * End the minigame and clean up resources.
   */
  public synchronized void end() {
    if (state == MinigameState.ENDED) {
      return;
    }
    setState(MinigameState.ENDED);
    try {
      onEnd();
    } catch (Exception e) {
      manager.getPlugin().getLogger().warning("Error during minigame end: " + e.getMessage());
    }
    for (Consumer<Minigame> listener : new ArrayList<>(endListeners)) {
      try {
        listener.accept(this);
      } catch (Exception e) {
        manager.getPlugin().getLogger().warning("Error in end listener: " + e.getMessage());
      }
    }

    //HandlerList.unregisterAll(this);


    // Clear caches and collections to prevent memory leaks
    playerCache.clear();
    endListeners.clear();
    players.clear();
    // Loop logic: schedule restart if enabled
    if (shouldLoop) {
      manager.getPlugin().getServer().getScheduler().runTaskLater(
          manager.getPlugin(),
          this::restartMinigame,
          loopDelayTicks
      );
    }
  }

  /**
   * Restart the minigame for looping.
   * This resets state and calls initialize/start.
   */
  private void restartMinigame() {
    initialize();
    start();
  }

  /**
   * Add a player to the minigame.
   *
   * @param player Player to add
   * @return true if player was successfully added
   */
  public boolean addPlayer(@NotNull Player player) {
    Objects.requireNonNull(player, "Player cannot be null");
    synchronized (players) {
      if (players.contains(player.getUniqueId())) {
        return false;
      }
      if (players.size() >= maxPlayers) {
        player.sendMessage(MinigameConfig.getMsgGameFull());
        return false;
      }
      if (state == MinigameState.RUNNING && !allowJoinDuringGame) {
        player.sendMessage(MinigameConfig.getMsgGameInProgress());
        return false;
      }
      players.add(player.getUniqueId());
      playerCache.put(player.getUniqueId(), player);
      spawnPoint.ifPresent(player::teleport);
      try {
        onPlayerJoin(player);
      } catch (Exception e) {
        manager.getPlugin().getLogger().warning("Error in onPlayerJoin for " + player.getName() + ": " + e.getMessage());
      }
      if (state == MinigameState.WAITING && players.size() >= minPlayers) {
        startCountdown();
      }
      return true;
    }
  }

  /**
   * Remove a player from the minigame.
   *
   * @param player Player to remove
   */
  public void removePlayer(@NotNull Player player) {
    Objects.requireNonNull(player, "Player cannot be null");
    synchronized (players) {
      if (!players.contains(player.getUniqueId())) {
        return;
      }
      players.remove(player.getUniqueId());
      onPlayerCleanup(player);
      onPlayerLeave(player);
      if (state == MinigameState.RUNNING && players.size() < minPlayers) {
        broadcastMessage(Component.text(MinigameConfig.getMsgPlayersRemaining()));
        end();
      }
      if (state == MinigameState.COUNTDOWN && players.size() < minPlayers) {
        setState(MinigameState.WAITING);
        broadcastMessage(Component.text(MinigameConfig.getMsgCountdownCancelled()));
      }
    }
  }

  /**
   * Send a message to all players in the minigame.
   *
   * @param message Message to send
   */
  public void broadcastMessage(@NotNull Component message) {
    Objects.requireNonNull(message, "Message cannot be null");
    synchronized (players) {
      for (UUID uuid : players) {
        Player player = getPlayerById(uuid);
        if (player != null) {
          player.sendMessage(message);
        }
      }
    }
  }

  /**
   * Start the countdown to begin the game.
   */
  public void startCountdown() {
    setState(MinigameState.COUNTDOWN);
    onCountdownStart();
  }

  /**
   * Add a listener that will be called when this minigame ends.
   *
   * @param listener Consumer that takes the ended minigame
   */
  public void addEndListener(@NotNull Consumer<Minigame> listener) {
    Objects.requireNonNull(listener, "Listener cannot be null");
    endListeners.add(listener);
  }

  /**
   * Set whether this minigame should automatically loop (restart) after ending.
   *
   * @param shouldLoop true to enable looping
   */
  public void setShouldLoop(boolean shouldLoop) {
    this.shouldLoop = shouldLoop;
  }

  /**
   * Set the delay (in ticks) before restarting the minigame if looping is enabled.
   *
   * @param delayTicks Delay in server ticks (20 ticks = 1 second)
   */
  public void setLoopDelayTicks(long delayTicks) {
    this.loopDelayTicks = delayTicks;
  }

  /**
   * Get whether this minigame is set to loop.
   */
  public boolean isShouldLoop() {
    return shouldLoop;
  }

  /**
   * Get the loop delay in ticks.
   */
  public long getLoopDelayTicks() {
    return loopDelayTicks;
  }

  /**
   * Get a player by UUID with caching for performance.
   *
   * @param uuid Player UUID
   * @return Player instance or null if not online
   */
  protected @Nullable Player getPlayerById(@NotNull UUID uuid) {
    Player cached = playerCache.get(uuid);
    if (cached != null && cached.isOnline()) {
      return cached;
    }

    Player player = manager.getPlugin().getServer().getPlayer(uuid);
    if (player != null) {
      playerCache.put(uuid, player);
    } else {
      playerCache.remove(uuid);
    }
    return player;
  }

  /**
   * Validate if a player is in a valid state for this minigame.
   *
   * @param player Player to validate
   * @return true if player is valid and can participate
   */
  protected boolean isPlayerValid(@Nullable Player player) {
    return player != null && player.isOnline() && players.contains(player.getUniqueId());
  }

  /**
   * Clean up expired entries from player cache.
   * Should be called periodically to prevent memory leaks.
   */
  public void cleanupPlayerCache() {
    playerCache.entrySet().removeIf(entry -> {
      Player player = entry.getValue();
      boolean shouldRemove = player == null || !player.isOnline() || !this.players.contains(entry.getKey());

      if (shouldRemove && player != null) {
        this.getManager().getPlugin().getLogger().fine(
            "Cleaning up cached player: " + player.getName() + " from minigame " + this.getId()
        );
      }
      return shouldRemove;
    });
  }

  protected void refreshAllOnlinePlayers() {
    for (UUID uuid : new HashSet<>(this.players)) {
      Player player = this.getPlayerById(uuid);
      if (player == null || !player.isOnline()) {
        // Player is offline, remove them
        this.players.remove(uuid);
        this.playerCache.remove(uuid);
      } else {
        // Refresh cache
        this.playerCache.put(uuid, player);
      }
    }
  }

  protected void onStateChange(MinigameState oldState, MinigameState newState) {
    this.getManager().getPlugin().getLogger().fine(
        "Minigame " + this.getId() + " state changed from " + oldState + " to " + newState
    );
  }

  public List<Player> getValidOnlinePlayers() {
    List<Player> validPlayers = new ArrayList<>();

    for (UUID uuid : new HashSet<>(this.players)) {
      Player player = this.getPlayerById(uuid);
      if (player != null && player.isOnline()) {
        validPlayers.add(player);
      } else {
        // Remove invalid players
        this.players.remove(uuid);
        this.playerCache.remove(uuid);
      }
    }

    return validPlayers;
  }

  public void validateGameState() {
    // Remove offline players
    this.refreshAllOnlinePlayers();

    // Check if game should end due to player count
    if (this.state == MinigameState.RUNNING && this.players.size() < this.minPlayers) {
      this.getManager().getPlugin().getLogger().warning(
          "Minigame " + this.getId() + " has too few players (" + this.players.size() +
              "/" + this.minPlayers + "), ending game"
      );
      this.end();
    }
  }

  public void performHealthCheck() {
    this.validateGameState();
    this.cleanupPlayerCache();
  }

  // Abstract methods that minigames must implement

  /**
   * Called when the minigame starts.
   * Override this method and call super.onStart() to maintain base functionality.
   */
  protected void onStart() {
    getManager().getPlugin().getLogger().info("Minigame " + getId() + " started with " + getPlayerCount() + " players");
  }

  /**
   * Called when the minigame ends.
   * Override this method and call super.onEnd() to maintain base functionality.
   */
  protected void onEnd() {
    cleanupPlayerCache();
    refreshAllOnlinePlayers();

    getManager().getPlugin().getLogger().info("Minigame " + getId() + " ended");
  }

  /**
   * Called when a player joins the minigame.
   * Override this method and call super.onPlayerJoin(player) to maintain base functionality.
   *
   * @param player Player who joined
   */
  protected void onPlayerJoin(@NotNull Player player) {
    playerCache.put(player.getUniqueId(), player);
    getManager().getPlugin().getLogger().info("Player " + player.getName() + " joined minigame " + getId());
  }

  /**
   * Called when a player leaves the minigame.
   * Override this method and call super.onPlayerLeave(player) to maintain base functionality.
   *
   * @param player Player who left
   */
  protected void onPlayerLeave(@NotNull Player player) {
    onPlayerCleanup(player);
    getManager().getPlugin().getLogger().info("Player " + player.getName() + " left minigame " + getId());
  }


  protected void onPlayerCleanup(Player player) {
    playerCache.remove(player.getUniqueId());
  }

  /**
   * Called when the countdown starts.
   * Override this method and call super.onCountdownStart() to maintain base functionality.
   */
  protected void onCountdownStart() {
    getManager().getPlugin().getLogger().info("Countdown started for minigame " + getId());
  }

  // Event handlers

  @EventHandler
  public void onPlayerQuitServer(PlayerQuitEvent event) {
    Player player = event.getPlayer();
    if (this.players.contains(player.getUniqueId())) {
      this.getManager().getPlugin().getLogger().info(
          "Player " + player.getName() + " quit server while in minigame " + this.getId()
      );

      // Schedule removal on next tick to avoid concurrent modification
      Bukkit.getScheduler().runTask(this.getManager().getPlugin(), () -> {
        this.removePlayer(player);
      });
    }
  }

  // Getters and Setters

  public @NotNull String getId() {
    return id;
  }

  public @NotNull String getDisplayName() {
    return displayName;
  }

  public @NotNull MinigameManager getManager() {
    return manager;
  }

  public @NotNull Set<UUID> getPlayers() {
    return Collections.unmodifiableSet(players);
  }

  public int getPlayerCount() {
    return players.size();
  }

  public @NotNull MinigameState getState() {
    return state;
  }


  public void setState(@NotNull MinigameState state) {
    MinigameState oldState = this.state;
    this.state = Objects.requireNonNull(state, "State cannot be null");

    // Call state change hook
    this.onStateChange(oldState, state);
  }

  public @NotNull Optional<Location> getSpawnPoint() {
    return spawnPoint;
  }

  public void setSpawnPoint(@Nullable Location spawnPoint) {
    this.spawnPoint = Optional.ofNullable(spawnPoint);
  }

  public int getMinPlayers() {
    return minPlayers;
  }

  public void setMinPlayers(int minPlayers) {
    if (minPlayers < 1) {
      throw new IllegalArgumentException("Minimum players must be at least 1");
    }
    if (minPlayers > maxPlayers) {
      throw new IllegalArgumentException("Minimum players cannot exceed maximum players");
    }
    this.minPlayers = minPlayers;
  }

  public int getMaxPlayers() {
    return maxPlayers;
  }

  public void setMaxPlayers(int maxPlayers) {
    if (maxPlayers < 1) {
      throw new IllegalArgumentException("Maximum players must be at least 1");
    }
    if (maxPlayers < minPlayers) {
      throw new IllegalArgumentException("Maximum players cannot be less than minimum players");
    }
    this.maxPlayers = maxPlayers;
  }

  protected boolean shouldAutoRemoveOnQuit(Player player) {
    return true; // Default behavior - can be overridden
  }

  protected boolean shouldAutoStart() {
    return true;
  }

  public boolean isAllowJoinDuringGame() {
    return allowJoinDuringGame;
  }

  public void setAllowJoinDuringGame(boolean allowJoinDuringGame) {
    this.allowJoinDuringGame = allowJoinDuringGame;
  }
}

File: src/main/kotlin/org/alpacaindustries/iremiaminigamecore/minigame/Minigame.kt
package org.alpacaindustries.iremiaminigamecore.minigame

import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import net.kyori.adventure.text.Component
import org.bukkit.event.EventHandler
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

/**
 * Base abstract class for all minigames.
 * Provides the core functionality and lifecycle management for minigames.
 *
 * This class is thread-safe for all public methods unless otherwise noted.
 */
abstract class Minigame(
    val id: String,
    val displayName: String,
    val manager: MinigameManager
) : Listener {

    private val _players = Collections.synchronizedSet(mutableSetOf<UUID>())
    private val endListeners = Collections.synchronizedList(mutableListOf<Consumer<Minigame>>())
    private val playerCache = ConcurrentHashMap<UUID, Player>()
    private val isInitialized = AtomicBoolean(false)
    private val isDestroyed = AtomicBoolean(false)

    @Volatile
    var spawnPoint: Location? = null

    @Volatile
    var state: MinigameState = MinigameState.WAITING
        private set

    @Volatile
    var minPlayers: Int = MinigameConfig.getDefaultMinPlayers()
        set(value) {
            require(value >= 1) { "Minimum players must be at least 1" }
            require(value <= maxPlayers) { "Minimum players cannot exceed maximum players" }
            field = value
        }

    @Volatile
    var maxPlayers: Int = MinigameConfig.getDefaultMaxPlayers()
        set(value) {
            require(value >= 1) { "Maximum players must be at least 1" }
            require(value >= minPlayers) { "Maximum players cannot be less than minimum players" }
            field = value
        }

    @Volatile
    var isAllowJoinDuringGame: Boolean = false

    var shouldLoop: Boolean = false
        private set

    var loopDelayTicks: Long = 60L // Default: 3 seconds (20 ticks = 1 second)
        private set

    init {
        require(id.trim().isNotEmpty()) { "Minigame ID cannot be empty" }
        require(displayName.trim().isNotEmpty()) { "Display name cannot be empty" }

        // Validate initial configuration
        require(minPlayers <= maxPlayers) { "Minimum players ($minPlayers) cannot exceed maximum players ($maxPlayers)" }
    }

    /**
     * Get an unmodifiable view of the players
     */
    val players: Set<UUID>
        get() = Collections.unmodifiableSet(_players)

    /**
     * Get the current player count
     */
    val playerCount: Int
        get() = _players.size

    /**
     * Check if the minigame has been properly initialized
     */
    val initialized: Boolean
        get() = isInitialized.get()

    /**
     * Check if the minigame has been destroyed
     */
    val destroyed: Boolean
        get() = isDestroyed.get()

    /**
     * Initialize the minigame. Registers event listeners and sets initial state.
     * This method is idempotent - calling it multiple times has no additional effect.
     */
    open fun initialize() {
        if (!isInitialized.compareAndSet(false, true)) {
            manager.plugin.logger.warning("Minigame $id is already initialized")
            return
        }

        if (isDestroyed.get()) {
            throw IllegalStateException("Cannot initialize destroyed minigame $id")
        }

        setState(MinigameState.WAITING)
        manager.plugin.server.pluginManager.registerEvents(this, manager.plugin)
        manager.plugin.logger.fine("Minigame $id initialized successfully")
    }

    /**
     * Start the minigame. Only allowed from WAITING or COUNTDOWN state.
     */
    @Synchronized
    open fun start() {
        checkNotDestroyed()

        if (state != MinigameState.WAITING && state != MinigameState.COUNTDOWN) {
            manager.plugin.logger.warning("Cannot start minigame $id from state $state")
            return
        }

        if (_players.size < minPlayers) {
            broadcastMessage(Component.text(MinigameConfig.getMsgNotEnoughPlayers()))
            return
        }

        setState(MinigameState.RUNNING)
        onStart()
    }

    /**
     * End the minigame and clean up resources.
     */
    @Synchronized
    open fun end() {
        if (state == MinigameState.ENDED) {
            return
        }

        setState(MinigameState.ENDED)

        try {
            onEnd()
        } catch (e: Exception) {
            manager.plugin.logger.warning("Error during minigame end: ${e.message}")
            e.printStackTrace()
        }

        // Notify all end listeners with error handling
        val listenersSnapshot = endListeners.toList()
        listenersSnapshot.forEach { listener ->
            try {
                listener.accept(this)
            } catch (e: Exception) {
                manager.plugin.logger.warning("Error in end listener: ${e.message}")
                e.printStackTrace()
            }
        }

        // Clear caches and collections to prevent memory leaks
        cleanup()

        // Loop logic: schedule restart if enabled
        if (shouldLoop && !isDestroyed.get()) {
            scheduleRestart()
        }
    }

    /**
     * Destroy the minigame and free all resources.
     * This is irreversible - the minigame cannot be used after destruction.
     */
    fun destroy() {
        if (!isDestroyed.compareAndSet(false, true)) {
            return // Already destroyed
        }

        try {
            // End the game if it's still running
            if (state != MinigameState.ENDED) {
                end()
            }

            // Unregister event listeners
            HandlerList.unregisterAll(this)

            // Final cleanup
            cleanup()

            manager.plugin.logger.info("Minigame $id destroyed successfully")
        } catch (e: Exception) {
            manager.plugin.logger.severe("Error destroying minigame $id: ${e.message}")
            e.printStackTrace()
        }
    }

    /**
     * Internal cleanup method
     */
    private fun cleanup() {
        playerCache.clear()
        endListeners.clear()
        // _players.clear()
    }

    /**
     * Schedule a restart for looping minigames
     */
    private fun scheduleRestart() {
        manager.plugin.server.scheduler.runTaskLater(
            manager.plugin,
            Runnable {
                if (!isDestroyed.get()) {
                    restartMinigame()
                }
            },
            loopDelayTicks
        )
    }

    /**
     * Restart the minigame for looping.
     * This resets state and calls initialize/start.
     */
    private fun restartMinigame() {
        try {
            if (isDestroyed.get()) {
                manager.plugin.logger.warning("Cannot restart destroyed minigame $id")
                return
            }

            // Reset initialization state for restart
            isInitialized.set(false)
            initialize()
            start()
        } catch (e: Exception) {
            manager.plugin.logger.severe("Error restarting minigame $id: ${e.message}")
            e.printStackTrace()
        }
    }

    /**
     * Add a player to the minigame.
     *
     * @param player Player to add
     * @return true if player was successfully added
     */
    open fun addPlayer(player: Player): Boolean {
        checkNotDestroyed()

        if (!player.isOnline) {
            return false
        }

        synchronized(_players) {
            if (_players.contains(player.uniqueId)) {
                return false
            }

            if (_players.size >= maxPlayers) {
                player.sendMessage(MinigameConfig.getMsgGameFull())
                return false
            }

            if (state == MinigameState.RUNNING && !isAllowJoinDuringGame) {
                player.sendMessage(MinigameConfig.getMsgGameInProgress())
                return false
            }

            _players.add(player.uniqueId)
            playerCache[player.uniqueId] = player
            spawnPoint?.let {
                try {
                    player.teleport(it)
                } catch (e: Exception) {
                    manager.plugin.logger.warning("Failed to teleport ${player.name} to spawn point: ${e.message}")
                }
            }

            try {
                onPlayerJoin(player)
            } catch (e: Exception) {
                manager.plugin.logger.warning("Error in onPlayerJoin for ${player.name}: ${e.message}")
                e.printStackTrace()
            }

            if (state == MinigameState.WAITING && _players.size >= minPlayers && shouldAutoStart()) {
                startCountdown()
            }
            return true
        }
    }

    /**
     * Remove a player from the minigame.
     *
     * @param player Player to remove
     */
    open fun removePlayer(player: Player) {
        synchronized(_players) {
            if (!_players.contains(player.uniqueId)) {
                return
            }

            _players.remove(player.uniqueId)

            try {
                onPlayerCleanup(player)
                onPlayerLeave(player)
            } catch (e: Exception) {
                manager.plugin.logger.warning("Error during player leave for ${player.name}: ${e.message}")
                e.printStackTrace()
            }

            // Check game state after player removal
            if (!isDestroyed.get()) {
                checkGameStateAfterPlayerLeave()
            }
        }
    }

    /**
     * Check game state after a player leaves and take appropriate action
     */
    private fun checkGameStateAfterPlayerLeave() {
        when (state) {
            MinigameState.RUNNING -> {
                if (_players.size < minPlayers) {
                    broadcastMessage(Component.text(MinigameConfig.getMsgPlayersRemaining()))
                    end()
                }
            }
            MinigameState.COUNTDOWN -> {
                if (_players.size < minPlayers) {
                    setState(MinigameState.WAITING)
                    broadcastMessage(Component.text(MinigameConfig.getMsgCountdownCancelled()))
                }
            }
            else -> { /* No action needed for other states */ }
        }
    }

    /**
     * Send a message to all players in the minigame.
     *
     * @param message Message to send
     */
    fun broadcastMessage(message: Component) {
        val playersSnapshot = _players.toSet()
        playersSnapshot.forEach { uuid ->
            try {
                getPlayerById(uuid)?.sendMessage(message)
            } catch (e: Exception) {
                manager.plugin.logger.fine("Failed to send message to player $uuid: ${e.message}")
            }
        }
    }

    /**
     * Start the countdown to begin the game.
     */
    open fun startCountdown() {
        checkNotDestroyed()
        setState(MinigameState.COUNTDOWN)
        onCountdownStart()
    }

    /**
     * Add a listener that will be called when this minigame ends.
     *
     * @param listener Consumer that takes the ended minigame
     */
    fun addEndListener(listener: Consumer<Minigame>) {
        checkNotDestroyed()
        endListeners.add(listener)
    }

    /**
     * Remove an end listener.
     *
     * @param listener Consumer to remove
     * @return true if the listener was removed
     */
    fun removeEndListener(listener: Consumer<Minigame>): Boolean {
        return endListeners.remove(listener)
    }

    /**
     * Check if the minigame has been destroyed and throw if it has
     */
    private fun checkNotDestroyed() {
        if (isDestroyed.get()) {
            throw IllegalStateException("Minigame $id has been destroyed")
        }
    }

    /**
     * Set whether this minigame should automatically loop (restart) after ending.
     *
     * @param shouldLoop true to enable looping
     */
    fun setShouldLoop(shouldLoop: Boolean) {
        checkNotDestroyed()
        this.shouldLoop = shouldLoop
    }

    /**
     * Set the delay (in ticks) before restarting the minigame if looping is enabled.
     *
     * @param delayTicks Delay in server ticks (20 ticks = 1 second)
     */
    fun setLoopDelayTicks(delayTicks: Long) {
        checkNotDestroyed()
        require(delayTicks > 0) { "Loop delay must be positive" }
        this.loopDelayTicks = delayTicks
    }

    /**
     * Get a player by UUID with caching for performance.
     * This method is thread-safe and includes automatic cache cleanup.
     *
     * @param uuid Player UUID
     * @return Player instance or null if not online
     */
    protected fun getPlayerById(uuid: UUID): Player? {
        // Fast path: check cache first
        val cached = playerCache[uuid]
        if (cached?.isOnline == true) {
            return cached
        }

        // Slow path: query server and update cache
        val player = manager.plugin.server.getPlayer(uuid)
        return if (player?.isOnline == true) {
            playerCache[uuid] = player
            player
        } else {
            // Clean up stale cache entry
            playerCache.remove(uuid)
            null
        }
    }

    /**
     * Validate if a player is in a valid state for this minigame.
     *
     * @param player Player to validate
     * @return true if player is valid and can participate
     */
    protected fun isPlayerValid(player: Player?): Boolean {
        return player?.isOnline == true && _players.contains(player.uniqueId)
    }

    /**
     * Clean up expired entries from player cache.
     * Should be called periodically to prevent memory leaks.
     * This method is optimized to minimize performance impact.
     */
    fun cleanupPlayerCache() {
        if (playerCache.isEmpty()) return

        val toRemove = mutableListOf<UUID>()

        playerCache.forEach { (uuid, player) ->
            if (!player.isOnline || !_players.contains(uuid)) {
                toRemove.add(uuid)
            }
        }

        toRemove.forEach { uuid ->
            playerCache.remove(uuid)
            manager.plugin.logger.fine("Cleaned up cached player: $uuid from minigame $id")
        }
    }

    /**
     * Refresh all online players and remove offline ones.
     * This method ensures data consistency between player list and actual online status.
     */
    protected fun refreshAllOnlinePlayers() {
        val toRemove = mutableSetOf<UUID>()

        _players.forEach { uuid ->
            val player = getPlayerById(uuid)
            if (player?.isOnline != true) {
                toRemove.add(uuid)
            } else {
                playerCache[uuid] = player
            }
        }

        // Remove offline players
        toRemove.forEach { uuid ->
            _players.remove(uuid)
            playerCache.remove(uuid)
        }

        if (toRemove.isNotEmpty()) {
            manager.plugin.logger.fine("Removed ${toRemove.size} offline players from minigame $id")
        }
    }

    /**
     * Handle state changes with proper logging and validation.
     */
    protected open fun onStateChange(oldState: MinigameState, newState: MinigameState) {
        manager.plugin.logger.fine("Minigame $id state changed from $oldState to $newState")

        // Validate state transitions
        when (oldState to newState) {
            MinigameState.WAITING to MinigameState.COUNTDOWN,
            MinigameState.COUNTDOWN to MinigameState.RUNNING,
            MinigameState.COUNTDOWN to MinigameState.WAITING,
            MinigameState.RUNNING to MinigameState.ENDED -> { /* Valid transitions */ }
            MinigameState.WAITING to MinigameState.RUNNING -> { /* Direct start - valid */ }
            else -> {
                if (newState != MinigameState.ENDED) { // Ending is always valid
                    manager.plugin.logger.warning("Unusual state transition in minigame $id: $oldState -> $newState")
                }
            }
        }
    }

    /**
     * Get all valid online players with automatic cleanup.
     * This method is thread-safe and performs inline cleanup of offline players.
     */
    fun getValidOnlinePlayers(): List<Player> {
        val validPlayers = mutableListOf<Player>()
        val toRemove = mutableSetOf<UUID>()

        _players.forEach { uuid ->
            val player = getPlayerById(uuid)
            if (player?.isOnline == true) {
                validPlayers.add(player)
            } else {
                toRemove.add(uuid)
            }
        }

        // Clean up offline players
        toRemove.forEach { uuid ->
            _players.remove(uuid)
            playerCache.remove(uuid)
        }

        return validPlayers
    }

    /**
     * Validate the current game state and take corrective action if needed.
     * This includes checking player counts and online status.
     */
    fun validateGameState() {
        if (isDestroyed.get()) return

        refreshAllOnlinePlayers()

        if (state == MinigameState.RUNNING && _players.size < minPlayers) {
            manager.plugin.logger.warning(
                "Minigame $id has too few players (${_players.size}/$minPlayers), ending game"
            )
            end()
        }
    }

    /**
     * Perform a comprehensive health check of the minigame.
     * This should be called periodically by the manager.
     */
    fun performHealthCheck() {
        if (isDestroyed.get()) return

        try {
            validateGameState()
            cleanupPlayerCache()

            // Log health status for debugging
            manager.plugin.logger.fine(
                "Health check for minigame $id: State=$state, Players=${_players.size}/$maxPlayers, Cache=${playerCache.size}"
            )
        } catch (e: Exception) {
            manager.plugin.logger.warning("Error during health check for minigame $id: ${e.message}")
            e.printStackTrace()
        }
    }

    /**
     * Set the minigame state and trigger state change hook.
     * This method ensures proper state transition validation.
     */
    protected fun setState(newState: MinigameState) {
        val oldState = this.state
        this.state = newState
        onStateChange(oldState, newState)
    }

    /**
     * Get detailed status information about this minigame.
     * Useful for debugging and monitoring.
     */
    fun getStatusInfo(): String {
        return buildString {
            append("Minigame[$id]: ")
            append("State=$state, ")
            append("Players=${_players.size}/$maxPlayers, ")
            append("Initialized=${isInitialized.get()}, ")
            append("Destroyed=${isDestroyed.get()}, ")
            append("Loop=$shouldLoop")
        }
    }

    // Abstract methods that minigames must implement

    /**
     * Called when the minigame starts.
     * Override this method and call super.onStart() to maintain base functionality.
     */
    protected open fun onStart() {
        manager.plugin.logger.info("Minigame $id started with $playerCount players")
    }

    /**
     * Called when the minigame ends.
     * Override this method and call super.onEnd() to maintain base functionality.
     */
    protected open fun onEnd() {
        cleanupPlayerCache()
        refreshAllOnlinePlayers()
        manager.plugin.logger.info("Minigame $id ended")
    }

    /**
     * Called when a player joins the minigame.
     * Override this method and call super.onPlayerJoin(player) to maintain base functionality.
     *
     * @param player Player who joined
     */
    protected open fun onPlayerJoin(player: Player) {
        playerCache[player.uniqueId] = player
        manager.plugin.logger.info("Player ${player.name} joined minigame $id")
    }

    /**
     * Called when a player leaves the minigame.
     * Override this method and call super.onPlayerLeave(player) to maintain base functionality.
     *
     * @param player Player who left
     */
    protected open fun onPlayerLeave(player: Player) {
        onPlayerCleanup(player)
        manager.plugin.logger.info("Player ${player.name} left minigame $id")
    }

    /**
     * Called to clean up player-specific resources when they leave.
     * Override this method to add custom cleanup logic.
     *
     * @param player Player being cleaned up
     */
    protected open fun onPlayerCleanup(player: Player) {
        playerCache.remove(player.uniqueId)
    }

    /**
     * Called when the countdown starts.
     * Override this method and call super.onCountdownStart() to maintain base functionality.
     */
    protected open fun onCountdownStart() {
        manager.plugin.logger.info("Countdown started for minigame $id")
    }


    // Event handlers - import missing dependencies

    @EventHandler
    fun onPlayerQuitServer(event: org.bukkit.event.player.PlayerQuitEvent) {
        val player = event.player
        if (_players.contains(player.uniqueId)) {
            manager.plugin.logger.info(
                "Player ${player.name} quit server while in minigame $id"
            )

            // Schedule removal on next tick to avoid concurrent modification
            org.bukkit.Bukkit.getScheduler().runTask(manager.plugin, Runnable {
                removePlayer(player)
            })
        }
    }

    /**
     * Determines if players should be automatically removed when they quit the server.
     * Override this method to customize auto-removal behavior.
     *
     * @param player Player who quit
     * @return true if player should be automatically removed
     */
    protected open fun shouldAutoRemoveOnQuit(player: Player): Boolean = true

    /**
     * Determines if the minigame should automatically start when enough players join.
     * Override this method to customize auto-start behavior.
     *
     * @return true if minigame should auto-start
     */
    protected open fun shouldAutoStart(): Boolean = true
}
